### ISSUE TITLE
Question about the implementation of VAMPIRE

### ISSUE THREAD

**Issue by AyiStar** [cnguyen10/few_shot_meta_learning#17]
Hi Cuong,

I really appreciate your work for the VAMPIRE algorithm. There are some questions about the implementation in Vampire2.py. 

1. Why is the KL loss in the inner loop implemented as the KL distribution between q(\theta) and the standard Gaussion, instead of KL(q||p)? I do not find the correspondence in the original paper of VAMPIRE.
2. Why does the global update (i.e., validation() function in the code) also need a KL loss?

Thank you.

**Comment by cnguyen10** [comment]
Hi @AyiStar,

Yes, you are right that the implementation is different from the one mentioned in the paper. Originally, it was exactly the same  and implemented in `Vampire.py` (you can trace back to git log to see the file since I deleted from the git repo). Then when I understood more about meta-learning, I thought that would be more reasonable to consider meta-learning as an extension of hyper-parameter optimization. In that case, it is analogous to train - validation split and the objective is a bi-level optimization. That is why I change the implementation to `Vampire2.py`.

For further details, please refer to chapter 3 in my [thesis](https://digital.library.adelaide.edu.au/dspace/handle/2440/135234). Let me know if you still have any questions regarding to the implementation.

**Comment by AyiStar** [comment]
@cnguyen10 Thank you for your reply, but I still do not totally get the two questions I asked. and I beg your further illustration.

1. I have read Chapter 3 of your thesis, and I do not understand why the KL divergence loss in the thesis (Eq. 3.10 in the thesis) discards the dependence on the meta-parameters. That is also the first question in my prior comment.
2. If you add the KL loss in the global update step as in Vampire2.py, how that corresponds to the original objective for the maximum predictive distribution (Eq. 3.5 in the thesis) ?

**Comment by cnguyen10** [comment]
I forgot to mention about the section 3.1 (especially Eq. (3.2)) in chapter 3 since the remaining is just a copy of the published paper.

As I said earlier, meta-learning is an extension of hyper-parameter where the hyper-parameter of interest (called meta-parameter) is shared across all tasks. In hyper-parameter optimization, we use the train - validation split to find a local optimal hyper-parameter through the following objective:

$$
\begin{align}
& \min_{\theta} \frac{1}{m^{(v)}} \sum_{k = 1}^{m^{(v)}}  \ell(y_{k}^{(v)}, f(x_{k}^{(v)}; w^{\star}(\theta)) \\
& \text{s.t.: } w^{\star}(\theta) \in \arg\min_{w} \frac{1}{m^{(t)}} \sum_{j = 1}^{m^{(t)}} \ell(y_{j}, f(x_{j}^{(t)}; w(\theta)),
\end{align}
$$

where: $\ell$ is a loss function and $f(.; w)$ is a model parameterized by $w$.

In this case, the objective of meta-learning can be written as:

$$
\begin{align}
& \min_{\theta} \textcolor{Red}{\frac{1}{T} \sum_{i = 1}^{T}} \frac{1}{m_{i}^{(v)}} \sum_{k = 1}^{m_{i}^{(v)}}  \ell(y_{ik}^{(v)}, f(x_{ik}^{(v)}; w_{i}^{\star}(\theta)) \\
& \text{s.t.: } w_{i}^{\star}(\theta) \in \arg\min_{w_{i}} \frac{1}{m_{i}^{(t)}} \sum_{j = 1}^{m_{i}^{(t)}} \ell(y_{ij}, f(x_{ij}^{(t)}; w_{i}(\theta)) \textcolor{Red}{\forall i \in \{1, \ldots, T\}},
\end{align}
$$

Thus, by replacing the loss function $\ell$, we can obtain different meta-learning algorithm. For example, when $\ell$ is the negative log-likelihood, the meta-learning algorithm is MAML.

Given this unified framework, I modified the previously-proposed Vampire algorithm by replacing $\ell$ by the variational-free energy (negative ELBO). In this case, the loss is consistent between train and validation. That explains why in this new perspective, we have KL divergence on both the lower- and upper-levels of the optimization objective.

However, if you are still keen on the original implementation, you could simply replace the standard Gaussian prior $p(w)$ by $\mathcal{N}(w; \mu_{\theta}, \sigma^{2}_{\theta})$, and remove the KL term in the upper-level optimization. The results reported in the paper followed the original implementation presented in the paper.

**Comment by AyiStar** [comment]
Thanks for your detailed explanation !
